src/interpreter/mod.rs

```diff
@@ -726,7 +775,63 @@ where
         }
 
         //state empty implies that either the execution has terminated or we have a
-        //Pk based descriptor
+        //Pk based descriptor or a Covenant descriptor
+        if let Some(pk) = self.cov {
+            // First verify the top of Miniscript.
+            // At this point, the stack must contain 13 elements
+            // pop the satisfied top and verify the covenant code.
+            if self.stack.pop() != Some(stack::Element::Satisfied) {
+                return Some(Err(Error::IncorrectCovenantWitness));
+            }
+            if self.stack.len() != 12 {
+                return Some(Err(Error::UnexpectedStackEnd));
+            }
+            // safe to unwrap 12 times
+            for i in 0..12 {
+                if let Err(e) = self.stack[i].try_push() {
+                    return Some(Err(e));
+                }
+            }
+            let mut ser_sig = Vec::new();
+            // 1.29 errors
+            {
+                let sighash_bytes = self.stack[11].as_push();
+                let sighash_u32 = util::slice_to_u32_le(sighash_bytes);
+                let sighash_ty = SigHashType::from_u32(sighash_u32);
+                let sig_vec = self.stack[0].as_push();
+                ser_sig.extend(sig_vec);
+                ser_sig.push(sighash_ty as u8);
+            }
+
+            if let Ok(sig) = verify_sersig(&mut self.verify_sig, &pk, &ser_sig) {
+                //Signature check successful, set cov to None to
+                //terminate the next() function in the subsequent call
+                self.cov = None;
+                // Do the checkSigFromStackCheck
+                let sighash_msg: Vec<u8> = self.stack.0[1..]
+                    .into_iter()
+                    .map(|x| Vec::from(x.as_push()))
+                    .flatten()
+                    .collect();
+                let mut eng = SigHash::engine();
+                eng.input(&sighash_msg);
+                let sighash_u256 = SigHash::from_engine(eng);
+                let msg = bitcoin::secp256k1::Message::from_slice(&sighash_u256[..]).unwrap();
+
+                // TODO: THIS SHOULD BE A SEPARATE PARAMETER TO THE FUNCTION, BUT SINCE
+                // IT MIGHT ELSEWHERE, CONSIDER MAKING IT A SEPARATE METHOD. RIGHT NOW,
+                // THIS IS CREATING A NEW CONTEXT WHICH IS EXPENSIVE
+                let secp = secp256k1::Secp256k1::verification_only();
```

Comment:
This is rather annoying as an API change. Makes the API even more difficult to use if we allow passing `Option<secp256k1::Secp256k1<C>> where C: verification.

For now, I am creating a new context each time we interpret which is not ideal but works for testing.

Reply:
@apoelstra, any comments on this?

Reply:
So, to handle the existing checksigs we have the user pass in a closure. There were two reasons for this
* To check normal signatures you need to compute a sighash, which is hard to do from the interpreter
* I wanted to be able to interpret miniscripts without checking the signatrues, since this is really expensive and doesn't give you much value if you're checking things that are in the chain anyway

I have a couple thoughts about how we could handle this here
* Not check the signature and offer no way to do so (this seems like a bad idea)
* Make the user pass a second closure in for this (ughh)
* Adapt the existing closure to take a message hash (ugly, doesn't really match the existing closure signature)
* Replace the existing closure with an optional secp context argument (but then how can we compute the sighash for normal checksigs?)
* Require our signatures to have R = P = 1, and then we can verify the signature with `SecretKey::tweak_add_assign` :P

None of these are really clean, but I'm leaning toward adding a second closure to the API.

Reply:
We don't really need a second closure if we add an additional optional secp parameter. 

I partially implemented it this way, but it proved annoying to the type of C `Option<secp256k1::Secp256k1<C>> where C: Verification whenever we supplied None. 

If we make a non-optional one, it might be inefficient as the users would be required to create a secp context, even if they are not using covenants, but I think that should be cleaner if it is acceptable as a tradeoff.

Reply:
Taking a closure seems a bit excessive because there will only ever be 1 checksigfromstack in a descriptor.

Reply:
Punting on this for now as it's not important. Will create a reminder issue.